kv/kvnemesis: TestKVNemesisMultiNode_BufferedWritesNoPipelining failed

kv/kvnemesis.TestKVNemesisMultiNode_BufferedWritesNoPipelining [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20372153?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20372153?buildTab=artifacts#/) on release-25.3 @ [26d912c2d5d3307b71df5c1e37998c8b3a5bc3da](https://github.com/cockroachdb/cockroach/commits/26d912c2d5d3307b71df5c1e37998c8b3a5bc3da):


```
          txn.Put(ctx, tk(15069434931838912630), sv(31)) // omitted
          txn.ScanForShareGuaranteedDurability(ctx, tk(8106530427441869066), tk(13683085621665096429), 0) // omitted
          txn.Put(ctx, tk(7667741866797789949), sv(32)) // omitted
          return nil
        }) // kv/kvclient/kvcoord/dist_sender.go:1134: batch with SkipLocked wait policy contains Put request: kv/kvclient/kvcoord/dist_sender.go:1134: batch with SkipLocked wait policy contains Put request
    kvnemesis.go:173: error applying x.ScanForShareSkipLockedGuaranteedDurability(ctx, tk(1695705150765324074), tk(18241945724473461016), 0) // kv/kvclient/kvcoord/dist_sender.go:1134: batch with SkipLocked wait policy contains Put request: kv/kvclient/kvcoord/dist_sender.go:1134: batch with SkipLocked wait policy contains Put request
    kvnemesis.go:173: error applying x.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
          txn.SetIsoLevel(isolation.Snapshot)
          txn.SetBufferedWritesEnabled(true)
          txn.GetForUpdateSkipLockedGuaranteedDurability(ctx, tk(12613453335943274215)) // @1755750017.598012155,0 (<nil>, <nil>)
          txn.ReverseScanForShare(ctx, tk(4199913419428481444), tk(17436723869849992695), 0) // @1755750018.017531229,1 (/Table/100/"f0241ef671057cad":v12, /Table/100/"a2dcd74f3a936071":v27, /Table/100/"91f0aa1126f98e74":v27, /Table/100/"8acdb1fcb2da4da2":v27, /Table/100/"6a2ec83855617a57":v27, /Table/100/"6a1e9c79fa4d57c3":v27, /Table/100/"58b692488d0deee6":v27, /Table/100/"4dd0ca4ca18eceaf":v27, <nil>)
          {
            b := &kv.Batch{}
            b.GetForUpdate(tk(15036517401567645265)) // (<nil>, <nil>)
            txn.Run(ctx, b) // @1755750018.017531229,1 <nil>
          }
          txn.ScanForUpdateSkipLockedGuaranteedDurability(ctx, tk(2106205237835378576), tk(5116069013835372603), 0) // @1755750018.017531229,1 (/Table/100/"21a21fd4cf03cb87":v27, <nil>)
          txn.GetForUpdateSkipLockedGuaranteedDurability(ctx, tk(3169397635625861761)) // @1755750018.017531229,1 (<nil>, <nil>)
          txn.ReverseScanSkipLocked(ctx, tk(5651939751042268006), tk(7387249643425033497), 0) // @1755750018.017531229,1 (/Table/100/"58b692488d0deee6":v27, <nil>)
          txn.ScanForShareSkipLocked(ctx, tk(4696497560069277295), tk(15716419692175949988), 0) // @1755750018.017531229,1 (/Table/100/"4dd0ca4ca18eceaf":v27, /Table/100/"58b692488d0deee6":v27, /Table/100/"6a1e9c79fa4d57c3":v27, /Table/100/"6a2ec83855617a57":v27, /Table/100/"8acdb1fcb2da4da2":v27, /Table/100/"91f0aa1126f98e74":v27, /Table/100/"a2dcd74f3a936071":v27, <nil>)
          txn.Put(ctx, tk(8520782534851061687), sv(41)) // <nil>
          txn.ReverseScanForUpdateSkipLockedGuaranteedDurability(ctx, tk(387057826567654922), tk(7860200185749704560), 0) // kv/kvclient/kvcoord/dist_sender.go:1134: batch with SkipLocked wait policy contains Put request
          txn.ReverseScanForUpdateSkipLocked(ctx, tk(11090284278826500808), tk(12417188540468246012), 0) // omitted
          txn.ReverseScanForUpdate(ctx, tk(557850909188939369), tk(7700496791140072379), 0) // omitted
          return nil
        }) // kv/kvclient/kvcoord/dist_sender.go:1134: batch with SkipLocked wait policy contains Put request: kv/kvclient/kvcoord/dist_sender.go:1134: batch with SkipLocked wait policy contains Put request
    kvnemesis.go:173: error applying x.ReverseScanForUpdateSkipLockedGuaranteedDurability(ctx, tk(387057826567654922), tk(7860200185749704560), 0) // kv/kvclient/kvcoord/dist_sender.go:1134: batch with SkipLocked wait policy contains Put request: kv/kvclient/kvcoord/dist_sender.go:1134: batch with SkipLocked wait policy contains Put request
    kvnemesis.go:208: failures(verbose): /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis1162416849/failures
        repro steps: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis1162416849/repro.go
        rangefeed KVs: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis1162416849/kvs-rangefeed.txt
        scan KVs: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis1162416849/kvs-scan.txt
    kvnemesis_test.go:459: [0] proposed: 2308
    kvnemesis_test.go:461: [0] reproposed unchanged: 74
    kvnemesis_test.go:463: [0] reproposed with new LAI: 1
    kvnemesis_test.go:459: [1] proposed: 168
    kvnemesis_test.go:461: [1] reproposed unchanged: 0
    kvnemesis_test.go:463: [1] reproposed with new LAI: 6
    kvnemesis_test.go:459: [2] proposed: 0
    kvnemesis_test.go:461: [2] reproposed unchanged: 0
    kvnemesis_test.go:463: [2] reproposed with new LAI: 0
    kvnemesis_test.go:468: 
        	Error Trace:	pkg/kv/kvnemesis/kvnemesis_test.go:468
        	            				pkg/kv/kvnemesis/kvnemesis_test.go:363
        	Error:      	Should be zero, but was 10
        	Test:       	TestKVNemesisMultiNode_BufferedWritesNoPipelining
        	Messages:   	kvnemesis detected failures
    dump_raft_logs.go:32: dumping raft logs to /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/kvnemesis1162416849
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/4aeb9c412c16a003ed8ddc43788d2bba/logTestKVNemesisMultiNode_BufferedWritesNoPipelining2530408933
--- FAIL: TestKVNemesisMultiNode_BufferedWritesNoPipelining (19.16s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #151047 kv/kvnemesis: TestKVNemesisMultiNode_BufferedWritesNoPipelining failed [sysbyte &amp; syscount mismatch] [A-testing C-bug C-test-failure O-robot P-3 T-kv X-duplicate branch-master]
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode_BufferedWritesNoPipelining.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-53720

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go
@@ -46,7 +46,7 @@ var BufferedWritesEnabled = settings.RegisterBoolSetting(
 
 var unsupportedInProductionBuildErr = errors.New("this option is not supported in production builds")
 
-var bufferedWritesScanTransformEnabled = settings.RegisterBoolSetting(
+var BufferedWritesScanTransformEnabled = settings.RegisterBoolSetting(
 	settings.ApplicationLevel,
 	"kv.transaction.write_buffering.transformations.scans.enabled",
 	"if enabled, locking scans and reverse scans with replicated durability are transformed to unreplicated durability",
@@ -59,7 +59,7 @@ var bufferedWritesScanTransformEnabled = settings.RegisterBoolSetting(
 	}),
 )
 
-var bufferedWritesGetTransformEnabled = settings.RegisterBoolSetting(
+var BufferedWritesGetTransformEnabled = settings.RegisterBoolSetting(
 	settings.ApplicationLevel,
 	"kv.transaction.write_buffering.transformations.get.enabled",
 	"if enabled, locking get requests with replicated durability are transformed to unreplicated durability",
@@ -294,8 +294,8 @@ func (twb *txnWriteBuffer) SendLocked(
 	// We check if scan transforms are enabled once and use that answer until the
 	// end of SendLocked.
 	cfg := transformConfig{
-		transformScans: bufferedWritesScanTransformEnabled.Get(&twb.st.SV),
-		transformGets:  bufferedWritesGetTransformEnabled.Get(&twb.st.SV),
+		transformScans: BufferedWritesScanTransformEnabled.Get(&twb.st.SV),
+		transformGets:  BufferedWritesGetTransformEnabled.Get(&twb.st.SV),
 	}
 
 	if twb.batchRequiresFlush(ctx, ba, cfg) {

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_client_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_client_test.go
@@ -61,7 +61,7 @@ func TestTxnCoordSenderWriteBufferingDisablesPipelining(t *testing.T) {
 	// buffered.
 	require.NoError(t, db.Put(ctx, "test-key-a", "hello"))
 
-	bufferedWritesScanTransformEnabled.Override(ctx, &st.SV, false)
+	BufferedWritesScanTransformEnabled.Override(ctx, &st.SV, false)
 	BufferedWritesMaxBufferSize.Override(ctx, &st.SV, defaultBufferSize)
 
 	// Without write buffering
@@ -144,7 +144,7 @@ func TestTxnWriteBufferFlushedWithMaxKeysOnBatch(t *testing.T) {
 
 	// The bug requires that we transform Gets. Here, we disable it to prove that
 	// this tets passes.
-	bufferedWritesGetTransformEnabled.Override(ctx, &st.SV, false)
+	BufferedWritesGetTransformEnabled.Override(ctx, &st.SV, false)
 
 	// The locks need to actually be taken, so let's write to every key we are
 	// going to lock.

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go
@@ -31,8 +31,8 @@ func makeMockTxnWriteBuffer(
 ) (txnWriteBuffer, *mockLockedSender, *cluster.Settings) {
 	metrics := MakeTxnMetrics(time.Hour)
 	st := cluster.MakeClusterSettings()
-	bufferedWritesScanTransformEnabled.Override(ctx, &st.SV, true)
-	bufferedWritesGetTransformEnabled.Override(ctx, &st.SV, true)
+	BufferedWritesScanTransformEnabled.Override(ctx, &st.SV, true)
+	BufferedWritesGetTransformEnabled.Override(ctx, &st.SV, true)
 	BufferedWritesMaxBufferSize.Override(ctx, &st.SV, defaultBufferSize)
 
 	mockSender := &mockLockedSender{}
@@ -1733,7 +1733,7 @@ func TestTxnWriteBufferLimitsSizeOfScans(t *testing.T) {
 				txn := makeTxnProto()
 				txn.Sequence = 10
 
-				bufferedWritesScanTransformEnabled.Override(ctx, &st.SV, true)
+				BufferedWritesScanTransformEnabled.Override(ctx, &st.SV, true)
 				BufferedWritesMaxBufferSize.Override(ctx, &st.SV, tc.bufferSize)
 
 				ba := &kvpb.BatchRequest{Header: kvpb.Header{Txn: &txn}}

--- a/pkg/kv/kvnemesis/kvnemesis_test.go
+++ b/pkg/kv/kvnemesis/kvnemesis_test.go
@@ -347,6 +347,10 @@ func TestKVNemesisMultiNode_BufferedWrites(t *testing.T) {
 		bufferedWriteProb:            0.70,
 		testSettings: func(ctx context.Context, st *cluster.Settings) {
 			kvcoord.BufferedWritesEnabled.Override(ctx, &st.SV, true)
+			// Read transforms are disabled on release-25.3 because of known issues
+			// such as #150239 which are only fixed in 25.4+.
+			kvcoord.BufferedWritesGetTransformEnabled.Override(ctx, &st.SV, false)
+			kvcoord.BufferedWritesScanTransformEnabled.Override(ctx, &st.SV, false)
 			concurrency.UnreplicatedLockReliabilityLeaseTransfer.Override(ctx, &st.SV, true)
 			concurrency.UnreplicatedLockReliabilityMerge.Override(ctx, &st.SV, true)
 			concurrency.UnreplicatedLockReliabilitySplit.Override(ctx, &st.SV, true)
@@ -372,6 +376,10 @@ func TestKVNemesisMultiNode_BufferedWritesNoPipelining(t *testing.T) {
 		testSettings: func(ctx context.Context, st *cluster.Settings) {
 			kvcoord.BufferedWritesEnabled.Override(ctx, &st.SV, true)
 			kvcoord.PipelinedWritesEnabled.Override(ctx, &st.SV, false)
+			// Read transforms are disabled on release-25.3 because of known issues
+			// such as #150239 which are only fixed in 25.4+.
+			kvcoord.BufferedWritesGetTransformEnabled.Override(ctx, &st.SV, false)
+			kvcoord.BufferedWritesScanTransformEnabled.Override(ctx, &st.SV, false)
 			concurrency.UnreplicatedLockReliabilityLeaseTransfer.Override(ctx, &st.SV, true)
 			concurrency.UnreplicatedLockReliabilityMerge.Override(ctx, &st.SV, true)
 			concurrency.UnreplicatedLockReliabilitySplit.Override(ctx, &st.SV, true)

